From 13a9969f21e519e68eb188dfdc250863ac729453 Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Fri, 7 Sep 2018 14:23:10 +0200
Subject: [PATCH 15/34] netlink: add netlink handler for sprivflags
 (--set-priv-flags)

Implement "ethtool --set-priv-flags <dev>" subcommand using netlink
interface command ETHTOOL_MSG_SETTINGS_SET.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 ethtool.c          |  3 ++-
 netlink/extapi.h   |  1 +
 netlink/settings.c | 35 +++++++++++++++++++++++++++++++++++
 3 files changed, 38 insertions(+), 1 deletion(-)

diff --git a/ethtool.c b/ethtool.c
index 2f23f7e6931d..6cbc42519db2 100644
--- a/ethtool.c
+++ b/ethtool.c
@@ -4984,6 +4984,7 @@ static int show_usage(struct cmd_context *ctx);
 #define nl_gfeatures	NULL
 #define nl_sfeatures	NULL
 #define nl_gprivflags	NULL
+#define nl_sprivflags	NULL
 #endif
 
 static const struct option {
@@ -5145,7 +5146,7 @@ static const struct option {
 	  "               [ combined N ]\n" },
 	{ "--show-priv-flags", 1, do_gprivflags, nl_gprivflags,
 	  "Query private flags" },
-	{ "--set-priv-flags", 1, do_sprivflags, NULL,
+	{ "--set-priv-flags", 1, do_sprivflags, nl_sprivflags,
 	  "Set private flags",
 	  "		FLAG on|off ...\n" },
 	{ "-m|--dump-module-eeprom|--module-info", 1, do_getmodule, NULL,
diff --git a/netlink/extapi.h b/netlink/extapi.h
index ce5fde1ddc35..f89c6d0e169c 100644
--- a/netlink/extapi.h
+++ b/netlink/extapi.h
@@ -20,6 +20,7 @@ int nl_sset(struct cmd_context *ctx);
 int nl_gfeatures(struct cmd_context *ctx);
 int nl_sfeatures(struct cmd_context *ctx);
 int nl_gprivflags(struct cmd_context *ctx);
+int nl_sprivflags(struct cmd_context *ctx);
 int nl_monitor(struct cmd_context *ctx);
 
 void monitor_usage();
diff --git a/netlink/settings.c b/netlink/settings.c
index 458dca26e05f..049db0e50bd8 100644
--- a/netlink/settings.c
+++ b/netlink/settings.c
@@ -1143,3 +1143,38 @@ int nl_sfeatures(struct cmd_context *ctx)
 		return 0;
 	return nlctx->exit_code ?: 92;
 }
+
+/* private flags */
+
+int nl_sprivflags(struct cmd_context *ctx)
+{
+	struct nl_context *nlctx = ctx->nlctx;
+	int ret;
+
+	nlctx->cmd = "--set-priv-flags";
+	nlctx->argp = ctx->argp;
+	nlctx->argc = ctx->argc;
+	nlctx->devname = ctx->devname;
+
+	ret = msg_init(nlctx, ETHTOOL_MSG_SETTINGS_SET,
+		       NLM_F_REQUEST | NLM_F_ACK);
+	if (ret < 0)
+		return 2;
+	if (ethnla_fill_header(nlctx, ETHTOOL_A_SETTINGS_HEADER, ctx->devname,
+			       0, 0, 0))
+		return -EMSGSIZE;
+
+	ret = nl_parse_bitset(nlctx, ETHTOOL_A_SETTINGS_PRIV_FLAGS, NULL,
+		              NULL);
+	if (ret < 0)
+		return -EINVAL;
+
+	ret = ethnl_sendmsg(nlctx);
+	if (ret < 0)
+		return 2;
+	ret = ethnl_process_reply(nlctx, nomsg_reply_cb);
+	if (ret == 0)
+		return 0;
+	else
+		return nlctx->exit_code ?: 1;
+}
-- 
2.22.0

